src/hotspot/share/services/mallocLimit.cpp

```diff
@@ -0,0 +1,222 @@
+  const char* _p;
+
+public:
+  ParserHelper(const char* s) : _s(s), _end(s + strlen(s)), _p(s) {}
+
+  bool eof() const { return _p >= _end; }
+
+  // Check if string at position matches a malloclimit_mode_t.
+  // Advance position on match.
+  bool match_mode_flag(MallocLimitMode* out) {
+    if (eof()) {
+      return false;
+    }
+    if (strncasecmp(_p, "oom", 3) == 0) {
+      *out = MallocLimitMode::trigger_oom;
+      _p += 3;
+      return true;
+    } else if (strncasecmp(_p, "fatal", 5) == 0) {
+      *out = MallocLimitMode::trigger_fatal;
+      _p += 5;
+      return true;
+    }
+    return false;
+  }
+
+  // Check if string at position matches a category name.
+  // Advances position on match.
+  bool match_category(MEMFLAGS* out) {
+    if (eof()) {
+      return false;
+    }
+    const char* end = strchr(_p, ':');
+    if (end == nullptr) {
+      end = _end;
+    }
+    stringStream ss;
+    ss.print("%.*s", (int)(end - _p), _p);
+    MEMFLAGS f = NMTUtil::string_to_flag(ss.base());
+    if (f != mtNone) {
+      *out = f;
+      _p = end;
+      return true;
+    }
+    return false;
+  }
+
+  // Check if string at position matches a memory size (e.g. "100", "100g" etc).
+  // Advances position on match.
+  bool match_size(size_t* out) {
+    if (!eof()) {
+      char* remainder = nullptr;
+      if (parse_integer<size_t>(_p, &remainder, out)) {
+        assert(remainder > _p && remainder <= _end, "sanity");
+        _p = remainder;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Check if char at pos matches c; return true and advance pos if so.
+  bool match_char(char c) {
+    if (!eof() && (*_p) == c) {
+      _p ++;
+      return true;
+    }
+    return false;
+  }
+};
+
+MallocLimitSet::MallocLimitSet() {
+  reset();
+}
+
+void MallocLimitSet::set_global_limit(size_t s, MallocLimitMode flag) {
+  _glob.sz = s; _glob.mode = flag;
+}
+
+void MallocLimitSet::set_category_limit(MEMFLAGS f, size_t s, MallocLimitMode flag) {
+  const int i = NMTUtil::flag_to_index(f);
+  _cat[i].sz = s; _cat[i].mode = flag;
+}
+
+void MallocLimitSet::reset() {
+  set_global_limit(0, MallocLimitMode::trigger_fatal);
+  _glob.sz = 0; _glob.mode = MallocLimitMode::trigger_fatal;
+  for (int i = 0; i < mt_number_of_types; i++) {
+    set_category_limit(NMTUtil::index_to_flag(i), 0, MallocLimitMode::trigger_fatal);
+  }
+}
+
+void MallocLimitSet::print_on(outputStream* st) const {
+  static const char* flagnames[] = { "fatal", "oom" };
+  if (_glob.sz > 0) {
+    st->print_cr("MallocLimit: total limit: " PROPERFMT " (%s)", PROPERFMTARGS(_glob.sz),
+              flagnames[(int)_glob.mode]);
+  } else {
+    for (int i = 0; i < mt_number_of_types; i++) {
+      if (_cat[i].sz > 0) {
+        st->print_cr("MallocLimit: category \"%s\" limit: " PROPERFMT " (%s)",
+          NMTUtil::flag_to_enum_name(NMTUtil::index_to_flag(i)),
+          PROPERFMTARGS(_cat[i].sz), flagnames[(int)_cat[i].mode]);
+      }
+    }
+  }
+}
```

Comment:
I am not crazy about:

`static const char* flagnames[] = { "fatal", "oom" };`

which requires that we position the strings in `flagnames` in a way that matches `MallocLimitMode` enum values:

```
enum class MallocLimitMode {
  trigger_fatal = 0,
  trigger_oom   = 1
};
```

Could we do instead:

```
void MallocLimitSet::print_on(outputStream* st) const {
  if (_glob.sz > 0) {
    st->print_cr("MallocLimit: total limit: " PROPERFMT " (%s)",
                 PROPERFMTARGS(_glob.sz), MallocLimitSet::limit_name(_glob.mode));
  } else {
    for (int i = 0; i < mt_number_of_types; i++) {
      if (_cat[i].sz > 0) {
        st->print_cr("MallocLimit: category \"%s\" limit: " PROPERFMT " (%s)",
          NMTUtil::flag_to_enum_name(NMTUtil::index_to_flag(i)),
          PROPERFMTARGS(_cat[i].sz), MallocLimitSet::limit_name(_cat[i].mode));
      }
    }
  }
}

```
and add new API to `MallocLimitSet`:

`  static const char* const limit_name(MallocLimitMode l)  { return (l==MallocLimitMode::trigger_fatal)?"fatal":"oom"; }
`

This way we also avoid the `(int)` casting from the `enum`.

We could also potentially use `MallocLimitSet::limit_name()` in `match_mode_flag()` instead of hardcoding "oom" and "fatal"

Reply:
Okay, minus the exposure of the API. Its not needed outside of this file.